exercises/rational-numbers/src/example.c

```diff
@@ -0,0 +1,125 @@
+#include "rational_numbers.h"
+#include <math.h>
+
+static int16_t absolute_int(int16_t n)
+{
+   const int16_t mask = n >> 15;
+   return (n + mask) ^ mask;
+}
+
+static float nth_root(float x, float n)
+{
+   return pow(x, 1. / n);
+}
+
+static int16_t greatestCommonDivisor(int n, int m)
+{
+   int16_t remainder;
+   while (n != 0) {
+      remainder = m % n;
+      m = n;
+      n = remainder;
+   }
+   return m;
+}
+
+rational_t reduce(rational_t r)
+{
+   if (r.numerator == 0)        // shortcut rational of 0
+      return (rational_t) {
+      0, 1};
+
+   int16_t n = greatestCommonDivisor(r.numerator, r.denominator);
+   int16_t numerator = r.numerator / n;
+   int16_t denominator = r.denominator / n;
+
+   if (denominator < 0) {       // reduce sign
+      denominator = -denominator;
+      numerator = -numerator;
+   }
+
+   return (rational_t) {
+   numerator, denominator};
+}
+
+rational_t add(rational_t r1, rational_t r2)
+{
+   // r1 + r2 = a1/b1 + a2/b2 = (a1 * b2 + a2 * b1) / (b1 * b2)
+   int16_t numerator =
+       r1.numerator * r2.denominator + r2.numerator * r1.denominator;
+   int16_t denominator = 1;
+   if (numerator != 0)
+      denominator = r1.denominator * r2.denominator;
+   return (rational_t) {
```

Comment:
For correctness, I'd assume that `add()`, `subtract()`, `multiply()` and `divide()` should all be fed through `reduce()` before returning, or is that not right?

I'd assume that 2/6 + 2/6 = 2/3, not 4/6.

Reply:
Hmm. It did occur to me but tests pass as is so I left it. 
IIRC the example implementation is not shown to users; so I guess there's somewhat of a philosophical question here - what is the purpose of this implementation?

Is it to be a canonical/reference implementation? 
Or just something that ensures the tests are constructed correctly?

Reply:
> Or just something that ensures the tests are constructed correctly?

This! This file is just to make CI pass and to prove that an implementation is _possible_. It's not meant to be the _perfect_ solution :)

Reply:
> > Or just something that ensures the tests are constructed correctly?
> 
> This! This file is just to make CI pass and to prove that an implementation is _possible_. It's not meant to be the _perfect_ solution :)

Counterpoint: I've seen many students refer to the repository to look for a "good" example solution so I try to strive for an example that can serve as a good reference solution. It may not be optimal in runtime or concision, but it should serve to instruct.

Reply:
In this case I think it's an issue with the exercise definition itself. It would be nice if it required rationals to be in reduced form and the tests verified this.

Reply:
> Counterpoint: I've seen many students refer to the repository to look for a "good" example solution so I try to strive for an example that can serve as a good reference solution. It may not be optimal in runtime or concision, but it should serve to instruct.

I think it's always fine to make a good example solution (we try it too), but I don't think it's a good counterpoint. It should be stated, in mentoring, explicitly, that the examples are just that: example solutions. Not good solutions. Students shouldn't be looking at them at all. I actively encourage mentors to discourage students from looking at them, and adding that they are not canonical / idiomatic examples makes that _easier_. The better they are, the harder it is "not to cheat" (for some people).

> In this case I think it's an issue with the exercise definition itself. It would be nice if it required rationals to be in reduced form and the tests verified this.

Absolutely agreed. I would suggest adding track-specific tests for this!

Reply:
Ai. Good enough is always better than those with obvious problems 👍